### PR TITLE
Add CSV data loader that randomly selects rows.

### PR DESCRIPTION
**Pull request by whitingjr**
This pull request adds another Step implementation. Allowing host and uri information to be loaded from a CSV file.
 Each request will use the values in a row. The row being randomly selected.

### CODE REVIEW

File: core/src/main/java/io/hyperfoil/core/generators/RandomCsvRowStep.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+      public ColumnsBuilder columns() {
+         return new ColumnsBuilder();
+      }
+
+      public Builder file(String file) {
+         this.file = file;
+         return this;
+      }
+
+      public Builder headers(String hasHeader) {
```
Comment (rvansa) [inline core/src/main/java/io/hyperfoil/core/generators/RandomCsvRowStep.java]: Please make the parameter boolean. Parser will do the conversion automatically.

File: core/src/main/java/io/hyperfoil/core/generators/RandomCsvRowStep.java [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+         this.isHeaderRemoved = Boolean.parseBoolean(hasHeader);
+         return this;
+      }
+
+      public Builder removeQuotes(String isQuotesRemoved) {
```
Comment (rvansa) [inline core/src/main/java/io/hyperfoil/core/generators/RandomCsvRowStep.java]: Same as above

File: core/src/main/java/io/hyperfoil/core/generators/RandomCsvRowStep.java [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+            rows = reader.lines()
+                  .filter(comments)
+                  .map(s -> isQuotesRemoved ? s.replaceAll("\"","") : s)
+                  .map(line -> line.split(","))
+                  .collect(Collectors.toList());
+         } catch (IOException ioe) {
+            throw new BenchmarkDefinitionException(ioe.getMessage());
+         }
+         if (rows.isEmpty()) {
+            throw new BenchmarkDataException("Missing CSV row data. Rows were not detected after initial processing of file.");
```
Comment (rvansa) [inline core/src/main/java/io/hyperfoil/core/generators/RandomCsvRowStep.java]: Do we need a separate `BenchmarkDataException`? IMO using `BenchmarkDefinitionException` would be sufficient.

File: core/src/main/java/io/hyperfoil/core/generators/RandomCsvRowStep.java [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+   }
+
+   @Override
+   public void reserve(Session session) {
+      Arrays.asList(columnVars).forEach(var -> var.declareObject(session));
+   }
+
+   public static class Builder extends BaseStepBuilder {
+      private String file;
+      private Boolean isHeaderRemoved;
```
Comment (rvansa) [inline core/src/main/java/io/hyperfoil/core/generators/RandomCsvRowStep.java]: primitive `boolean` please

File: core/src/main/java/io/hyperfoil/core/generators/RandomCsvRowStep.java [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+      @Override
+      public List<Step> build(SerializableSupplier<Sequence> sequence) {
+         File f = new File(file);
+         if ( !f.exists() ) {
+            throw new BenchmarkDefinitionException("Supplied file cannot be found on system");
+         }
+         List<String[]> rows;
+         try ( BufferedReader reader = Files.newBufferedReader(Paths.get(f.getAbsolutePath())) ){
+            Predicate<String> comments = s -> (isHeaderRemoved ? !(s.trim().startsWith("#")) : true );
```
Comment (rvansa) [inline core/src/main/java/io/hyperfoil/core/generators/RandomCsvRowStep.java]: You assume that 'header' is anything that starts with `#`. If you intend to skip any such line I would rather call the property `skipComments`.

When writing `headers: true` I don't really understand what should that do with the headers (include/skip?); If I see `skipHeaders: true` then I'd think that it should ignore first non-empty line.

File: core/src/main/java/io/hyperfoil/core/generators/RandomCsvRowStep.java [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+                  .map(s -> isQuotesRemoved ? s.replaceAll("\"","") : s)
+                  .map(line -> line.split(","))
+                  .collect(Collectors.toList());
+         } catch (IOException ioe) {
+            throw new BenchmarkDefinitionException(ioe.getMessage());
+         }
+         if (rows.isEmpty()) {
+            throw new BenchmarkDataException("Missing CSV row data. Rows were not detected after initial processing of file.");
+         }
+         rows.forEach(arr -> {for (int i = 0 ; i < arr.length; i += 1) {
```
Comment (rvansa) [inline core/src/main/java/io/hyperfoil/core/generators/RandomCsvRowStep.java]: Newline after `{` please

File: core/src/test/resources/scenarios/random-csv.hf.yaml [hunk 7]
```diff
@@ -0,0 +1,31 @@
+name: simple benchmark
+http:
+  - host: http://localhost:8099
```
Comment (rvansa) [inline core/src/test/resources/scenarios/random-csv.hf.yaml]: You can remove the indent here

File: core/src/main/java/io/hyperfoil/core/generators/RandomCsvRowStep.java (hunk 1 above)
Comment (whitingjr) [reply core/src/main/java/io/hyperfoil/core/generators/RandomCsvRowStep.java]: Happy to.

File: core/src/main/java/io/hyperfoil/core/generators/RandomCsvRowStep.java (hunk 2 above)
Comment (whitingjr) [reply core/src/main/java/io/hyperfoil/core/generators/RandomCsvRowStep.java]: ok

File: core/src/main/java/io/hyperfoil/core/generators/RandomCsvRowStep.java (hunk 3 above)
Comment (whitingjr) [reply core/src/main/java/io/hyperfoil/core/generators/RandomCsvRowStep.java]: ok, I was undecided at the time.

File: core/src/main/java/io/hyperfoil/core/generators/RandomCsvRowStep.java (hunk 4 above)
Comment (whitingjr) [reply core/src/main/java/io/hyperfoil/core/generators/RandomCsvRowStep.java]: ok

File: core/src/main/java/io/hyperfoil/core/generators/RandomCsvRowStep.java (hunk 5 above)
Comment (whitingjr) [reply core/src/main/java/io/hyperfoil/core/generators/RandomCsvRowStep.java]: ok

File: core/src/main/java/io/hyperfoil/core/generators/RandomCsvRowStep.java (hunk 6 above)
Comment (whitingjr) [reply core/src/main/java/io/hyperfoil/core/generators/RandomCsvRowStep.java]: ok

File: core/src/test/resources/scenarios/random-csv.hf.yaml (hunk 7 above)
Comment (whitingjr) [reply core/src/test/resources/scenarios/random-csv.hf.yaml]: ok

**Comment by whitingjr** [comment]
Branch has been updated to follow your comments.

File: core/src/main/java/io/hyperfoil/core/generators/RandomCsvRowStep.java [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+      public ColumnsBuilder columns() {
+         return new ColumnsBuilder();
+      }
+
+      public Builder file(String file) {
+         this.file = file;
+         return this;
+      }
+
+      public Builder headers(boolean hasHeader) {
```
Comment (rvansa) [inline core/src/main/java/io/hyperfoil/core/generators/RandomCsvRowStep.java]: please call this method `skipComments` and update the example yaml

**Comment by whitingjr** [comment]
I intended not to include a testcase. There was an issue that I had attempted to recreate but a fix was put in already.

File: core/src/main/java/io/hyperfoil/core/generators/RandomCsvRowStep.java (hunk 8 above)
Comment (whitingjr) [reply core/src/main/java/io/hyperfoil/core/generators/RandomCsvRowStep.java]: done

**Comment by rvansa** [comment]
Okay, merged. Well without a test case to run it the examples sit sort of abandoned, but let's keep them if you wish.

**Comment by whitingjr** [comment]
I'll see if i can hunt down the old testcase.

**Comment by rvansa** [comment]
Feel free to just push it to master (if if works).

**Pull request merged**